### ISSUE TITLE
can't pickle function objects with SGEJobTask

### ISSUE THREAD

**Issue by mmendez12** [spotify/luigi#1714]
I've been using luigi.Task to run jobs on a SGE queuing system without any problem but when I try the example from the SGEJobTask documentation I get the following error:

```
Traceback (most recent call last):
  File "/home/toto/anaconda2/lib/python2.7/site-packages/luigi-2.1.1-py2.7.egg/luigi/worker.py", l
ine 191, in run
    new_deps = self._run_get_new_deps()
  File "/home/toto/anaconda2/lib/python2.7/site-packages/luigi-2.1.1-py2.7.egg/luigi/worker.py", l
ine 129, in _run_get_new_deps
    task_gen = self.task.run()
  File "/home/toto/anaconda2/lib/python2.7/site-packages/luigi-2.1.1-py2.7.egg/luigi/contrib/sge.p
y", line 214, in run
    self._init_local()
  File "/home/toto/anaconda2/lib/python2.7/site-packages/luigi-2.1.1-py2.7.egg/luigi/contrib/sge.p
y", line 205, in _init_local
    self._dump(self.tmp_dir)
  File "/home/toto/anaconda2/lib/python2.7/site-packages/luigi-2.1.1-py2.7.egg/luigi/contrib/sge.p
y", line 232, in _dump
    d = pickle.dumps(self)
  File "/home/toto/anaconda2/lib/python2.7/copy_reg.py", line 70, in _reduce_ex
    raise TypeError, "can't pickle %s objects" % base.__name__
TypeError: can't pickle function objects
```

Here is the code I tried to run on a cluster, it is not exactly what's on the documentation but both code returned the same error above:

``` python
import luigi
from luigi.contrib.sge import SGEJobTask

class TestJobTask(SGEJobTask):

    i = luigi.Parameter()

    def work(self):

        with open(self.output().path, 'w') as f:
            f.write('this is a test')

    def output(self):
        return luigi.LocalTarget('./testfile_' + str(self.i))

if __name__ == '__main__':
    tasks = [TestJobTask(i=str(i), shared_tmp_dir='./') for i in range(3)]
    luigi.build(tasks, local_scheduler=True, workers=3)
```

I checked the source code but I am not familiar with pickle so I am not sure what is going wrong here.

Does anyone know how to fix this? maybe @jfeala ?

**Comment by Tarrasch** [comment]
Hmm, maybe you can @-ping the authors of the SGE module?

**Comment by mmendez12** [comment]
Thanks for the suggestion @Tarrasch I added it to the issue.

**Comment by jfeala** [comment]
Hi @mmendez12, glad you're using this module! What happens if you use absolute paths? The `shared_tmp_dir` is passed from the master to the worker node, so the workers won't understand a relative path.

**Comment by mmendez12** [comment]
Hi @jfeala I tried using absolute paths but I got exactly the same error... Any idea about what I could try next?

**Comment by real1991** [comment]
I am facing exactly the same problem when I work with Luigi and PysparkTask...

**Comment by mmendez12** [comment]
@jfeala, @real1991 I noticed the following condition right before pickeling the current object (sge.py l.231, spark.py l.288): `if self.__module__ == '__main__':` 

I am running the script with the following command `python my_workflow.py --local-server` instead of using the luigi command line interface, could that be an Issue here? @real1991 do you run your script the same way?

**Comment by real1991** [comment]
I have also tried the command 
`python SimpleTest.py TestTaskPS --local-scheduler` but it returns error message when running "d = pickle.dumps(self)" (spark.py I.289), saying "TypeError: can't pickle function objects". 

My code is similar with yours since I only would like to try a simple example:
```
import luigi
from luigi.contrib.spark import PySparkTask, SparkSubmitTask

class TestTaskPS(PySparkTask):
    def input(self):
        return luigi.LocalTarget('data/dat_0.json')

    def output(self):
        return luigi.LocalTarget('out_data3/')

    def main(self, sc, *args):
        lines = sc.textFile(self.input().path)
        lines.saveAsTextFile(self.output().path)
        
if __name__ == '__main__':
    luigi.run()
```

**Comment by mfcabrera** [comment]
Closing as it looks that this issues was fixed by #1765 and #1819
